File: 953-reverse-only-letters/reverse-only-letters.java
class Solution {
    public String reverseOnlyLetters(String s) {
        char[] chars = s.toCharArray();
        int start = 0, end = s.length() - 1;

        while(start < end) {
            if(!Character.isLetter(s.charAt(start))) {
                start++;
            } else if(!Character.isLetter(s.charAt(end))) {
                end--;
            } else {
                // swap characters
                char temp = chars[start];
                chars[start] = chars[end];
                chars[end] = temp;
                start++;
                end--;
            }
        }

        return new String(chars);
    }
}

File: 953-reverse-only-letters/reverse-only-letters.py
class Solution:
    def reverseOnlyLetters(self, s: str) -> str:
        left, right = 0, len(s) - 1

        c = list(s)

        while left < len(c) and left < right:
            while not c[left].isalpha() and left < right:
                print(f"{left}, {right}")
                left += 1
            while not c[right].isalpha() and right > left:
                print(f"{left}, {right}")
                right -= 1
            
            c[left], c[right] = c[right], c[left]
            left += 1
            right -= 1
        
        return "".join(c)